spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java

```diff
@@ -0,0 +1,267 @@
+import static org.apache.iceberg.TableProperties.PARQUET_VECTORIZATION_ENABLED;
+
+public class TestMergeIntoTable extends SparkRowLevelOperationsTestBase {
+  private final String sourceName;
+  private final String targetName;
+
+  public TestMergeIntoTable(String catalogName, String implementation, Map<String, String> config,
+                            String fileFormat, Boolean vectorized) {
+    super(catalogName, implementation, config, fileFormat, vectorized);
+    this.sourceName = tableName("source");
+    this.targetName = tableName("target");
+  }
+
+  @BeforeClass
+  public static void setupSparkConf() {
+    spark.conf().set("spark.sql.shuffle.partitions", "4");
+  }
+
+  protected Map<String, String> extraTableProperties() {
+    return ImmutableMap.of(TableProperties.DELETE_MODE, "copy-on-write");
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", targetName);
+    sql("DROP TABLE IF EXISTS %s", sourceName);
+  }
+
+  @Test
+  public void testEmptyTargetInsertAllNonMatchingRows() throws NoSuchTableException {
+    createAndInitUnPartitionedTargetTable(targetName);
+    createAndInitSourceTable(sourceName);
+    append(sourceName, new Employee(1, "emp-id-1"), new Employee(2, "emp-id-2"), new Employee(3, "emp-id-3"));
+    String sqlText = "MERGE INTO " + targetName + " AS target " +
+                     "USING " + sourceName + " AS source " +
+                     "ON target.id = source.id " +
+                     "WHEN NOT MATCHED THEN INSERT * ";
+
+    sql(sqlText);
+    sql("SELECT * FROM %s ORDER BY id, dep", targetName);
+    assertEquals("Should have expected rows",
+            ImmutableList.of(row(1, "emp-id-1"), row(2, "emp-id-2"), row(3, "emp-id-3")),
+            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+  }
+
+  @Test
+  public void testEmptyTargetInsertOnlyMatchingRows() throws NoSuchTableException {
+    createAndInitUnPartitionedTargetTable(targetName);
+    createAndInitSourceTable(sourceName);
+    append(sourceName, new Employee(1, "emp-id-1"), new Employee(2, "emp-id-2"), new Employee(3, "emp-id-3"));
+    String sqlText = "MERGE INTO " + targetName + " AS target " +
+                     "USING " + sourceName + " AS source " +
+                     "ON target.id = source.id " +
+                     "WHEN NOT MATCHED AND (source.id >= 2) THEN INSERT * ";
+
+    sql(sqlText);
+    List<Object[]> res = sql("SELECT * FROM %s ORDER BY id, dep", targetName);
+    assertEquals("Should have expected rows",
+            ImmutableList.of(row(2, "emp-id-2"), row(3, "emp-id-3")),
+            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+  }
+
+  @Test
+  public void testOnlyUpdate() throws NoSuchTableException {
+    createAndInitUnPartitionedTargetTable(targetName);
+    createAndInitSourceTable(sourceName);
+    append(targetName, new Employee(1, "emp-id-one"), new Employee(6, "emp-id-6"));
```

Comment:
The employee dep is identical for both records with id 6, so the assertion can't distinguish between the case where employee 6 is replaced or not. Could you update the original target data to `emp-id-six` and assert that it is unchanged because of the `target.id = 1` requirement?